# <span style="color: rgb(230, 0, 126);">NIVELL I</span>


Des d’un Jupyter Notebook faràs els següents exercicis utilitzant la llibreria requests de Python.

Exploració bàsica amb una API de laboratori.

**1.** Consulta l’API pública JSONPlaceholder utilitzant el mètode GET per obtenir:

- Llista de publicacions (/posts)

- Llista d’usuaris (/users)

- Llista de tasques (/todos)

Instalamos request e importamos la librería:

In [4]:
pip install requests

   ---------------------------------------- 0.0/64.7 kB ? eta -:--:--
   ------------------------------------- -- 61.4/64.7 kB 1.6 MB/s eta 0:00:01
   ---------------------------------------- 64.7/64.7 kB 865.0 kB/s eta 0:00:00
   ---------------------------------------- 0.0/152.9 kB ? eta -:--:--
   ------------------------------------- -- 143.4/152.9 kB 4.3 MB/s eta 0:00:01
   ---------------------------------------- 152.9/152.9 kB 2.3 MB/s eta 0:00:00
   ---------------------------------------- 0.0/107.0 kB ? eta -:--:--
   ---------------------------------------- 107.0/107.0 kB 3.1 MB/s eta 0:00:00
   ---------------------------------------- 0.0/71.0 kB ? eta -:--:--
   ---------------------------------------- 71.0/71.0 kB 2.0 MB/s eta 0:00:00
   ---------------------------------------- 0.0/131.6 kB ? eta -:--:--
   ------------------------------------- -- 122.9/131.6 kB 7.0 MB/s eta 0:00:01
   ---------------------------------------- 131.6/131.6 kB 1.9 MB/s eta 0:00:00
Note: you m


[notice] A new release of pip is available: 24.0 -> 25.3
[notice] To update, run: C:\Users\rubit\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:
import requests

Establecemos que la URL de la base de datos que vamos a consultar:

In [3]:
base_URL = "https://jsonplaceholder.typicode.com"

Ahora estableceremos qué parte concreta de la URL queremos consultar y guardamos la cadena de texto que recibimos en un json para que Python pueda "entenderlo".
Posteriormente hacemos iteración para que nos muestre el listado de lo que consultamos.

Lista de publicaciones (/posts)

In [4]:
respuesta_posts = requests.get(f"{base_URL}/posts")

posts = respuesta_posts.json()

for post in posts:
    print(post["title"])

sunt aut facere repellat provident occaecati excepturi optio reprehenderit
qui est esse
ea molestias quasi exercitationem repellat qui ipsa sit aut
eum et est occaecati
nesciunt quas odio
dolorem eum magni eos aperiam quia
magnam facilis autem
dolorem dolore est ipsam
nesciunt iure omnis dolorem tempora et accusantium
optio molestias id quia eum
et ea vero quia laudantium autem
in quibusdam tempore odit est dolorem
dolorum ut in voluptas mollitia et saepe quo animi
voluptatem eligendi optio
eveniet quod temporibus
sint suscipit perspiciatis velit dolorum rerum ipsa laboriosam odio
fugit voluptas sed molestias voluptatem provident
voluptate et itaque vero tempora molestiae
adipisci placeat illum aut reiciendis qui
doloribus ad provident suscipit at
asperiores ea ipsam voluptatibus modi minima quia sint
dolor sint quo a velit explicabo quia nam
maxime id vitae nihil numquam
autem hic labore sunt dolores incidunt
rem alias distinctio quo quis
est et quae odit qui non
quasi id et eos tenet

Lista de usuarios (/users)

In [5]:
respuesta_users = requests.get(f"{base_URL}/users")

users = respuesta_users.json()

for user in users:
     print(f'ID: {user["id"]} | Nombre: {user["name"]} | Username: {user["username"]}')

ID: 1 | Nombre: Leanne Graham | Username: Bret
ID: 2 | Nombre: Ervin Howell | Username: Antonette
ID: 3 | Nombre: Clementine Bauch | Username: Samantha
ID: 4 | Nombre: Patricia Lebsack | Username: Karianne
ID: 5 | Nombre: Chelsey Dietrich | Username: Kamren
ID: 6 | Nombre: Mrs. Dennis Schulist | Username: Leopoldo_Corkery
ID: 7 | Nombre: Kurtis Weissnat | Username: Elwyn.Skiles
ID: 8 | Nombre: Nicholas Runolfsdottir V | Username: Maxime_Nienow
ID: 9 | Nombre: Glenna Reichert | Username: Delphine
ID: 10 | Nombre: Clementina DuBuque | Username: Moriah.Stanton


Llista de tasques (/todos)

In [6]:
respuesta_todos = requests.get(f"{base_URL}/todos")

tareas = respuesta_todos.json()

for tarea in tareas:
    print(f'Tarea: {tarea["title"]} | Asignada: {tarea["userId"]} | Completada: {tarea["completed"]}')

Tarea: delectus aut autem | Asignada: 1 | Completada: False
Tarea: quis ut nam facilis et officia qui | Asignada: 1 | Completada: False
Tarea: fugiat veniam minus | Asignada: 1 | Completada: False
Tarea: et porro tempora | Asignada: 1 | Completada: True
Tarea: laboriosam mollitia et enim quasi adipisci quia provident illum | Asignada: 1 | Completada: False
Tarea: qui ullam ratione quibusdam voluptatem quia omnis | Asignada: 1 | Completada: False
Tarea: illo expedita consequatur quia in | Asignada: 1 | Completada: False
Tarea: quo adipisci enim quam ut ab | Asignada: 1 | Completada: True
Tarea: molestiae perspiciatis ipsa | Asignada: 1 | Completada: False
Tarea: illo est ratione doloremque quia maiores aut | Asignada: 1 | Completada: True
Tarea: vero rerum temporibus dolor | Asignada: 1 | Completada: True
Tarea: ipsa repellendus fugit nisi | Asignada: 1 | Completada: True
Tarea: et doloremque nulla | Asignada: 1 | Completada: False
Tarea: repellendus sunt dolores architecto voluptatum |

**2.** Mostra per pantalla:

La quantitat total de cada recurs.

El codi d'estat de cada petició.

Hacemos un print determinando a qué nombre corresponde cada bloque, contamos la cantidad de datos que contiene y verificamos el estado:

In [7]:
print("POSTS")
print("Total:", len(posts))
print("Status code:", respuesta_posts.status_code)
print()

print("USUARIOS")
print("Total:", len(users))
print("Status code:", respuesta_users.status_code)
print()

print("TAREAS")
print("Total:", len(tareas))
print("Status code:", respuesta_todos.status_code)
print()

POSTS
Total: 100
Status code: 200

USUARIOS
Total: 10
Status code: 200

TAREAS
Total: 200
Status code: 200



Hay un total de 100 posts, 10 usuarios y 200 tareas. Todos retornan un status code 200, que significa "OK", que la respuesta para la solicitud HTTP ha sido exitosa.

**3.**  Fes una petició a una publicació inexistent per obtenir un error 404 i mostra el codi d'estat rebut.

Hacemos petición a una URL que no existe y mosramos el status code:

In [8]:
response_404 = requests.get(f"{base_URL}/posts/99999")

print("Status code:", response_404.status_code)

Status code: 404


**4.**  Fes una petició POST per crear una nova publicació fictícia. Inclou el títol, el cos del missatge i un userId.

- Mostra la resposta JSON.

- Mostra el codi d'estat.

In [9]:
new_post = {
    "title": "Publicación fictícia de prueba",
    "body": "Este es el cuerpo de la publicación fictícia.",
    "userId": 1
}

respuesta_post = requests.post(f"{base_URL}/posts", json=new_post)

print("Respuesta JSON:")
print(respuesta_post.json())
print("Status code:", respuesta_post.status_code)

Respuesta JSON:
{'title': 'Publicación fictícia de prueba', 'body': 'Este es el cuerpo de la publicación fictícia.', 'userId': 1, 'id': 101}
Status code: 201


No guarda los datos realmente pero simula la respuesta de forma correcta. El status code 201 indica que la solicitud se cumplió  y el servidor creó un nuevo recurso.

**5.** Fes una petició PATCH per modificar parcialment una publicació existent.

Mostra la resposta JSON.

Mostra el codi d'estat.

Hacemos modificación del título del primer post: 

In [11]:
patch_data = {
    "title": "Título modificado con PATCH"
}

respuesta_patch = requests.patch(f"{base_URL}/posts/1", json=patch_data)

print("Respuesta JSON:")
print(respuesta_patch.json())
print("Status code:", respuesta_patch.status_code)

Respuesta JSON:
{'userId': 1, 'id': 1, 'title': 'Título modificado con PATCH', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
Status code: 200


**6.** Fes una petició DELETE sobre una publicació.

Mostra la resposta JSON.

Mostra el codi d'estat.

In [12]:
respuesta_delete = requests.delete(f"{base_URL}/posts/1")

print("Respuesta JSON:")
print(respuesta_delete.json())
print("Status code:", respuesta_delete.status_code)

Respuesta JSON:
{}
Status code: 200


Vemos que el post 1 aparece vacío

# <span style="color: rgb(230, 0, 126);">NIVELL II</span>

Interacció amb una API pública real

**1.**  Explora el repositori de Public APIs i tria una API que permeti fer peticions GET.

API elegida: Dog API (https://dog.ceo/dog-api/)

**2.** Llegeix la documentació de l'API:

Revisa la secció d’endpoints disponibles. En un markdown, apunta com a mínim dos endpoints diferents.


Endpoints seleccionados:

1.`https://dog.ceo/api/breeds/list/all`: Devuelve una lista con todas las razas de perro

2.`https://dog.ceo/api/breed/hound/images`: Devuelve una lista de imagenes de la raza HOUND

Comprova si ofereix filtres o paràmetres opcionals interessants. En un markdown, anota’ls.

La API permite filtrar por raza directamente en la URL: "/breed/{raza}/images" --> https://dog.ceo/api/breed/chow/images

Permite obtener un nombre aleatorio de imagenes si añadimos un número al final: "/images/random/3"

**3**. Fes una petició GET senzilla:

Mostra el codi d'estat de la resposta.

Imprimeix de forma clara alguns dels camps de la resposta JSON.

In [ ]:
url = "https://dog.ceo/api/breeds/list/all"
#Vamos a buscar los datos al servidor:
response = requests.get(url) 

#Comprobamos la conexión sea OK:
print(f"Status code: {response.status_code}")

Status code: 200


In [ ]:
# Pasamos a JSON
data = response.json()

# Imprimimos las primeras 5 razas
razas = data['message'] #Los datos útiles de esta API van dentro de la llave message
print("Algunas razas:")
print(list(razas.keys())[:5])

Algunas razas:
['affenpinscher', 'african', 'airedale', 'akita', 'appenzeller']


**4.** Converteix les dades a un DataFrame de pandas:

Mostra les primeres files del DataFrame.

In [16]:
import pandas as pd

In [17]:
url_detalle = "https://dog.ceo/api/breed/hound/list"
res_detalle = requests.get(url_detalle)
datos_hound = res_detalle.json()['message'] # Hacemos una lista de strings

In [18]:
# Creamos el DataFrame
df_dogs_hound = pd.DataFrame(datos_hound, columns=['Subrazas de Hound'])


df_dogs_hound.head()

Subrazas de Hound
0            afghan
1            basset
2             blood
3           english
4            ibizan